services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/allu/CableReportServiceITests.kt

```diff
@@ -45,19 +50,86 @@ class CableReportServiceITests {
     private lateinit var mockWebServer: MockWebServer
     private lateinit var service: CableReportService
 
+    private lateinit var authToken: String
+
     @BeforeEach
     fun setup() {
         mockWebServer = MockWebServer()
 
         val baseUrl = mockWebServer.url("/").toUrl().toString()
         val properties = AlluProperties(baseUrl, "fake_username", "any_password", 2)
         val webClient = webClientWithLargeBuffer(WebClient.builder())
+        authToken = createMockToken()
         service = CableReportService(webClient, properties)
+        service.authToken = authToken
+        service.authExpiration = Instant.now().plusSeconds(3600)
+    }
+
+    @Nested
+    inner class GetToken {
+        @Test
+        fun `doesn't renew token when it's not expired`() {
+            val token = service.getToken()
+
+            assertThat(mockWebServer.requestCount).isZero()
+            assertThat(token).isEqualTo(authToken)
+        }
+
+        @Test
+        fun `renews token when it's not yet set`() {
+            service.authToken = null
+            val mockToken = addStubbedLoginResponse()
+
+            val token = service.getToken()
+
+            assertThat(mockWebServer.takeRequest()).isValidLoginRequest()
+            assertThat(token).isEqualTo(mockToken)
+            assertThat(service.authToken).isEqualTo(mockToken)
+            assertThat(service.authExpiration!!.isAfter(Instant.now())).isTrue()
+        }
+
+        @Test
+        fun `renews token when it's expiration date is null`() {
+            service.authExpiration = null
+            val mockToken = addStubbedLoginResponse()
+
+            val token = service.getToken()
+
+            assertThat(mockWebServer.takeRequest()).isValidLoginRequest()
+            assertThat(token).isEqualTo(mockToken)
+            assertThat(service.authToken).isEqualTo(mockToken)
+            assertThat(service.authExpiration!!.isAfter(Instant.now())).isTrue()
+        }
+
+        @Test
+        fun `renews token when it's expired`() {
+            service.authExpiration = Instant.now().minusSeconds(6 * 60)
```

Comment:
Would a shorter time (any time in the past) work?

Reply:
Yes, thank you. I don't know why I was thinking about the 5-minute delay here. Replaced with 1 second in the past.